File: hello-dart/lib/server/message.dart
import 'dart:convert';
import 'protocol.dart';

class Message {
  final String command;
  final String? streamId;
  final int? offset;
  final int? length;
  final String? message;
  final String? status;
  final int? size;
  final String? streams;

  const Message({
    required this.command,
    this.streamId,
    this.offset,
    this.length,
    this.message,
    this.status,
    this.size,
    this.streams,
  });

  String toJson() {
    final map = <String, dynamic>{'command': command};
    if (streamId != null) map['streamId'] = streamId;
    if (offset != null) map['offset'] = offset;
    if (length != null) map['length'] = length;
    if (message != null) map['message'] = message;
    if (status != null) map['status'] = status;
    if (size != null) map['size'] = size;
    if (streams != null) map['streams'] = streams;
    return jsonEncode(map);
  }

  static Message parse(String json) {
    final data = jsonDecode(json) as Map<String, dynamic>;
    if (!data.containsKey('command')) {
      throw ArgumentError('Invalid message: missing command field');
    }
    return Message(
      command: data['command'] as String,
      streamId: data['streamId'] as String?,
      offset: data['offset'] as int?,
      length: data['length'] as int?,
      message: data['message'] as String?,
      status: data['status'] as String?,
      size: data['size'] as int?,
      streams: data['streams'] as String?,
    );
  }

  static (CommandInfo, Message) parseCommand(String json) {
    final msg = parse(json);
    final info = Protocol.lookup(msg.command);
    if (info == null) {
      throw ArgumentError('Unknown command: ${msg.command}');
    }
    return (info, msg);
  }

  // --- Factory methods ---

  factory Message.connected(String connId) =>
      Message(command: 'CONNECTED', streamId: connId);

  factory Message.created(String streamId) =>
      Message(command: 'CREATED', streamId: streamId);

  factory Message.completed(String streamId) =>
      Message(command: 'COMPLETED', streamId: streamId);

  factory Message.closed(String streamId) =>
      Message(command: 'CLOSED', streamId: streamId);

  factory Message.status(String streamId, String status, int size) =>
      Message(command: 'STATUS', streamId: streamId, status: status, size: size);

  factory Message.streamList(String streams) =>
      Message(command: 'STREAM_LIST', streams: streams);

  factory Message.error(String message) =>
      Message(command: 'ERROR', message: message);
}

File: hello-php/src/Server/Message.php
<?php

declare(strict_types=1);

namespace AudioStream\Server;

final class Message
{
    public function __construct(
        public readonly string $command,
        public readonly ?string $streamId = null,
        public readonly ?int $offset = null,
        public readonly ?int $length = null,
        public readonly ?string $message = null,
        public readonly ?string $status = null,
        public readonly ?int $size = null,
        public readonly ?string $streams = null,
    ) {}

    public function toJson(): string
    {
        $data = ['command' => $this->command];
        if ($this->streamId !== null) $data['streamId'] = $this->streamId;
        if ($this->offset !== null) $data['offset'] = $this->offset;
        if ($this->length !== null) $data['length'] = $this->length;
        if ($this->message !== null) $data['message'] = $this->message;
        if ($this->status !== null) $data['status'] = $this->status;
        if ($this->size !== null) $data['size'] = $this->size;
        if ($this->streams !== null) $data['streams'] = $this->streams;
        return json_encode($data) ?: '{}';
    }

    public static function parse(string $json): self
    {
        $data = json_decode($json, true);
        if (!is_array($data) || !isset($data['command'])) {
            throw new \InvalidArgumentException('Invalid message: missing command field');
        }
        return new self(
            command: $data['command'],
            streamId: $data['streamId'] ?? null,
            offset: isset($data['offset']) ? (int) $data['offset'] : null,
            length: isset($data['length']) ? (int) $data['length'] : null,
            message: $data['message'] ?? null,
            status: $data['status'] ?? null,
            size: isset($data['size']) ? (int) $data['size'] : null,
            streams: $data['streams'] ?? null,
        );
    }

    /**
     * @return array{CommandInfo, Message}
     */
    public static function parseCommand(string $json): array
    {
        $msg = self::parse($json);
        $info = Protocol::lookup($msg->command);
        if ($info === null) {
            throw new \InvalidArgumentException("Unknown command: {$msg->command}");
        }
        return [$info, $msg];
    }

    // --- Factory methods ---

    public static function connected(string $connId): self
    {
        return new self(command: 'CONNECTED', streamId: $connId);
    }

    public static function created(string $streamId): self
    {
        return new self(command: 'CREATED', streamId: $streamId);
    }

    public static function completed(string $streamId): self
    {
        return new self(command: 'COMPLETED', streamId: $streamId);
    }

    public static function closed(string $streamId): self
    {
        return new self(command: 'CLOSED', streamId: $streamId);
    }

    public static function status(string $streamId, string $status, int $size): self
    {
        return new self(command: 'STATUS', streamId: $streamId, status: $status, size: $size);
    }

    public static function streamList(string $streams): self
    {
        return new self(command: 'STREAM_LIST', streams: $streams);
    }

    public static function error(string $message): self
    {
        return new self(command: 'ERROR', message: $message);
    }
}
